### ISSUE TITLE
Control not released while sending multiple messages

### ISSUE THREAD

**Issue by mishas** [aaugustin/websockets#865]
**TL;DR**: Sending many messages in a loop (via a loop of `await ws.send(...)` or via passing an async generator to `send()` once), does not release control, and disconnects clients due to unanswered pings.

**The longer story**:
I have a server that generates large messages to send to clients. Sending those messages takes time (about a minute), both because it takes some time to generate the messages themselves (CPU-intensive), but also because the network is rather slow.

The problem that I've noticed, is that `await ws.send(...)` does not release control even once done.
I also tried to pass an async generator to send, and didn't get control back also (see example code below).

The outcome from that is that websocket's pings are not getting answered, and the client disconnects.

**Example code**:
Here's a simple server that sends 20 packets of data to the client. I use `time.sleep(1)` instead of `asyncio.sleep(1)` to simulate actual CPU-intensive processing.
```
import asyncio
import time
import websockets

async def handler(websocket, path):
    async for request in websocket:
        print("Got a request to send stuff")
        for x in range(20):
            print("Sending", x)
            await websocket.send(str(x))
            time.sleep(1)  # Simulate CPU-intensive operation
        print("Done sending")

async def heartbeat():
    while True:
        await asyncio.sleep(1)
        print("Heartbeat: ", time.time())

async def main():
    await websockets.serve(handler, "localhost", 6789)
    await heartbeat()

asyncio.run(main())
```

The output looked something like this:
```
Heartbeat:  1607622802.574514
Got a request to send stuff
Sending 0
Sending 1
<... No heartbeats here ...>
Sending 18
Sending 19
Done sending
Heartbeat:  1607622823.145667
Heartbeat:  1607622824.1507761
```
As you can see, there're no hearbeats at all while in the sending loop, even though we do `await` each `websocket.send()` individually.

Adding `await asyncio.sleep(0)` right after the send, as a patch, does somewhat fixes the problem. (I start to get heardbeats every few (~5) seconds)

I also tried passing an async iterator into `ws.send()` (since the documentation of [send](https://websockets.readthedocs.io/en/stable/api.html#websockets.protocol.WebSocketCommonProtocol.send) clearly states that in that case control will be released), but no luck. Here's the new `handler()` function:
```
async def handler(websocket, path):
    async def gen():
        for x in range(20):
            yield str(x)
            time.sleep(1)  # Simulate CPU-heavy operation

    async for request in websocket:
        print("Got a request to send stuff")
        await websocket.send(gen())
        print("Done sending")
```
This is actually worse, as I didn't have a good place to add an `await asyncio.sleep(0)` in.

I've tested with Websockets 8.1, and both python3.7 and python3.8.

**Comment by lgrahl** [comment]
I can't tell you why the heartbeat does not appear but it's definitely a problem to have a long-blocking operation on the event loop thread. You should offload this operation via `loop.run_in_executor`.

**Comment by mishas** [comment]
@lgrahl, I added the time.sleep just to make it easier to create a smaller piece of code.
My real code is fully asyncio, thus there's nothing I can offload, as I don't have any "heavy" action anyways, just a lot of code.
Furthermore, due to slowness of the network, and the amount of packets sent, most of the "blocking" seem to be due to network rather than CPU.
(In the actual code, the limits of packets are raised, and i'm sending packets of ~10MB at a time).

Thanks for the quick response, though 👍

**Comment by lgrahl** [comment]
The term *blocking* refers to the current thread. Thus, an asynchronous task that takes a long time when being awaited is, by definition, not a *blocking* call. Synchronously executed CPU intensive parts however are blocking the current thread and preventing the event loop from being executed and therefore are called *blocking*. This differentiation is vital when using event loops.

Can you provide a short full example with server and client that reproduces the behaviour you're seeing?

**Comment by mishas** [comment]
@lgrahl,

I understand what blocking means in this context.
My understanding of Asyncio is, that if I `await` something (anything?) enough times I can be sure that all other coroutines will have a chance to run (no matter how much I'm blocking where).

In the example above, you can remove the `time.sleep(1)` and instead make a loop of `range(2**20)` (instead of `range(20)`), and you'll get the same result (it'll also help to do that over a slow network) - as long as the only `await`s you have are for `ws.send()` nothing else will run.
This means that `ws.send()` for some reason starves every other coroutine.

I couldn't reproduce it with any of the more standard awaits. I tried `await asyncio.sleep(0)` or `await session.get(...)` (for aiohttp's session), and both let heartbeats work as intended..

As requested, I'm attaching simple code for a client, and for reference, also the server without any blocking below.

The client is the most simple client:
```
import asyncio
import websockets

async def main():
    async with websockets.connect("ws://localhost:6789") as websocket:
        await websocket.send("Send me stuff")
        async for resp in websocket:
            print("Received:", resp)

asyncio.run(main())
```

For completeness, here's the full server code without anything blocking:
```
import asyncio
import time
import websockets

async def handler(websocket, path):
    async for request in websocket:
        print("Got a request to send stuff")
        for x in range(2**20):
            if x % 10000 == 0:
                print("Sending", x)
            await websocket.send(str(x))
        print("Done sending")

async def heartbeat():
    while True:
        await asyncio.sleep(1)
        print("Heartbeat: ", time.time())

async def main():
    await websockets.serve(handler, "localhost", 6789)
    await heartbeat()

asyncio.run(main())
```

**Comment by lgrahl** [comment]
That is indeed awkward and should be investigated. Even if the consumer is faster and the flow control does not kick in, it should eventually yield back to the event loop, so something is odd here. /cc @aaugustin 

> My understanding of Asyncio is, that if I await something (anything?) enough times I can be sure that all other coroutines will have a chance to run (no matter how much I'm blocking where).

IIRC ECMAScript yields control back to the event loop with each `await` but Python does not. You can experimentally try this yourselves:

```python
import asyncio
import time

future = asyncio.Future()
future.set_result(True)

async def noop():
    while True:
        await future
        #await asyncio.sleep(0.0)  # Uncomment and the heartbeat will get a chance to run

async def heartbeat():
    while True:
        await asyncio.sleep(1.0)
        print("Heartbeat: ", time.time())

async def main():
    await asyncio.gather(heartbeat(), noop())

asyncio.run(main())
```

**Comment by dbrgn** [comment]
Hm, asyncio not yielding back to the scheduler every time await is called would be a problem if a resource is always ready. As far as I know, asyncio is purely cooperative multitasking and running tasks cannot be preempted.

I'm not sure if this is exactly the same issue, but Tokio recently added [cooperative task yielding](https://tokio.rs/blog/2020-04-preemption) to address this issue. Is there a smiliar way in asyncio that would allow adding yield points in the Websockets code at locations where it makes sense (e.g. send loops)?

**Comment by dbrgn** [comment]
It appears that in Python `sleep(0)` can be used to explicitly yield control to the event loop: https://github.com/python/asyncio/issues/284

(In JS it would be similar, `setTimeout(..., 0)`: 
https://stackoverflow.com/questions/64814478/how-can-a-javascript-async-function-explicitly-yield-control-at-a-specific-point)

**Comment by lgrahl** [comment]
It looks like explicitly yielding back to the event loop is the solution here. Perhaps this should be done on every `.drain`, @aaugustin?

**Comment by mishas** [comment]
@lgrahl (and /cc @aaugustin)

This is actually interesting. We can even make this example simpler by not involving `future`s:
```python
import asyncio
import time

async def no_wait():
    return 5

async def noop():
    while True:
        await no_wait()
        #await asyncio.sleep(0.0)  # Uncomment and the heartbeat will get a chance to run

async def heartbeat():
    while True:
        await asyncio.sleep(1.0)
        print("Heartbeat: ", time.time())

async def main():
    await asyncio.gather(heartbeat(), noop())

asyncio.run(main())
```

It seems like it'll be easier to understand the problem by looking into history:
The `await` mechanism was built over `yield from` in python. If we'll write the same implementation in a pseudo-python3.3 syntax, it'll be something like this:
```python
def no_wait():
    if False:
        yield None
    return 5

def noop():
    while True:
        yield from no_wait()
```
As you can see above, the `no_wait` generator never yields anything, which means the `yield from` on the last line will never  yield anything, which means we'll never release control.

This explains why some `await`s are never returning to the control.
It's actually a good question if there should be a standard way to do that (which is not `sleep(0)`), or if python should enforce all `await`s to return to Control.

In any case, I believe that in the specific `websockets` case, it'll make sense for `websockets` to indeed (as mentioned above), force it.

Thanks for looking at it so quickly!

(PS. Sorry for closing/reopening the ticket, mis-click :( )

**Comment by aaugustin** [comment]
Here's my theory:

- You're filling the OS send buffer with data that will take about 1 minute to send.
- Within the next 20 seconds, websockets sends a ping, which remains stuck behind the rest of the data for at least 40 seconds.
- After 20 seconds, websockets doesn't get a pong, because the ping is still stuck in the OS' send buffer.

So websockets times out.

This isn't a matter of yielding to the event loop. It's a matter of clogging the send buffer.

Set `ping_timeout` to `120` or `None` and you should be good.

----

I have a hard time understanding your expectations around a heartbeat when you block the event loop with a CPU intensive task or a `time.sleep`. I'm unable to make a constructive comment on that part.

Also I don't think yielding control to the event loop is part of the contract of any API in websockets. (That would likely be a mistake.)

**Comment by mishas** [comment]
@aaugustin, if your theory about filling OS buffer is correct - why does `await asyncio.sleep(0)` helps?

Please also see above that we've discovered that this happens with other coroutines as well.

Also - the `time.sleep(1)` is just to make the example simpler. See examples above where it happens without that as well.

**Comment by aaugustin** [comment]
I read further; since you could reproduce the issue without involving websockets at all, I'm still skeptical that websockets is at fault here.

Unfortunately the code that handles draining has gotten pretty complicated due to a lack of extensibility of asyncio in this area -- I filed bugs against Python and the answer was "copy paste the code you need in your library". So I don't really know how the code I copy pasted works.

**Comment by aaugustin** [comment]
Also, I still take issue with the statement that:
> the documentation of send clearly states that in that case control will be released

I'm not seeing this in the documentation. I don't remember ever making guarantees about yielding control to the event loop. The general theory is that websockets attempts to behave as closely to asyncio as possible, to minimize surprises.

If you can demonstrate that asyncio behaves a certain way with a plain TCP stream, then that websockets has a worse behavior, then I'll consider this as an issue.

**Comment by dbrgn** [comment]
> I'm still skeptical that websockets is at fault here.

From what I understand, the websockets library is not at fault here, but it could improve _certain_ use cases that may lead to starvation (due to the fact how asyncio is designed) by adding a few yield points (`await sleep(0)`) in code paths that are called repeatedly (e.g. after sending n chunks/bytes of data).

The `await sleep(0)` does not really have a functional impact per se, but it allows other async tasks to kick in.

**Comment by mishas** [comment]
@aaugustin ,

> I'm not seeing this in the documentation.

From https://websockets.readthedocs.io/en/stable/api.html#websockets.protocol.WebSocketCommonProtocol.send:
> Indeed, there only two situations where send() yields control to the event loop:
>
> 1. The write buffer is full. If you don’t want to wait until enough data is sent, your only alternative is to close the connection. close() will likely time out then abort the TCP connection.
> 2. **message is an asynchronous iterator**. Stopping in the middle of a fragmented message will cause a protocol error. Closing the connection has the same effect.

Regarding what happens with plain TCP stream: the `write()` method of asyncio's streams is NOT an async function, so - I wouldn't expect it to give up control either way.
With streams the code is:
```python
stream.write(b"something")
await stream.drain()
```
In the above example, `drain` does yield control.

**Comment by lgrahl** [comment]
> In the above example, `drain` does yield control.

I wrote an example with basic asyncio streams and came to a different conclusion. There is no guarantee that `drain` yields, so this problem (also) exists in asyncio's TCP stream reader/writer implementations.

```python
import asyncio
import time

data = b'\x00' * 1024

async def handler(reader, writer):
    print('Start sending')
    for x in range(2**20):
        writer.write(data)
        await writer.drain()
    print("Done sending")
    writer.close()
    await writer.wait_closed()

async def heartbeat():
    while True:
        await asyncio.sleep(1.0)
        print("Heartbeat: ", time.time())

async def main():
    await asyncio.start_server(handler, 'localhost', 6789)
    await heartbeat()

asyncio.run(main())
```

```python
import asyncio

async def main():
    reader, writer = await asyncio.open_connection('localhost', 6789)
    while True:
        data = await reader.read()
        if len(data) == 0:
            break
        print("Received:", len(data))

asyncio.run(main())
```

**Comment by mishas** [comment]
Hmm, @lgrahl,

Indeed in the very simple case, it seems to act similarly.
But in real life cases it's a bit different.

Here's the important part from the documentation for `drain()`:
> drain() blocks until the size of the buffer is drained down to the low watermark and writing can be resumed. When there is nothing to wait for, the drain() returns immediately.

What it says is that if the buffer is below the low watermark, it returns immediately (thus not yielding control). I agree that it's a bit weird and counterintuitive, but makes some sense that if the function is a noop, don't yield.

In the case of `ws.send()`, it ALWAYS sends something, and as such, I believe is more confusing.

I don't think it's a good idea to compare `ws.send()` to `stream.drain()` though. If anything, it makes sense to compare to `stream.write()`, but in this case, it's an asyncio vs non-asyncio functions...

**Comment by lgrahl** [comment]
> What it says is that if the buffer is below the low watermark, it returns immediately (thus not yielding control). I agree that it's a bit weird and counterintuitive, but makes some sense that if the function is a noop, don't yield.

The issue is that if the consumer (client) is reading faster than the producer (server), and the producer's TCP buffer is being emptied faster than the Python code can produce packets (which is very likely because Python is not fast), it leads to the exact same problem.

> I don't think it's a good idea to compare `ws.send()` to `stream.drain()` though. If anything, it makes sense to compare to `stream.write()`, but in this case, it's an asyncio vs non-asyncio functions...

AFAIK websockets does pretty much that underneath, so I think the comparison is accurate.

---

I did a quick performance check on the above server (which transfer 1 GiB):

No yielding: 4.04s
Yielding after each drain: 7.14s with `asyncio.sleep(0)`

So, there is a significant performance impact.

Since asyncio's TCP streams are not very cooperative, IMO it is reasonable to leave the websockets implementation as is. A word of warning might be a good idea.

This was a big "today I learned" for me and I've been using asyncio for many years. If your server suddenly locks up for several seconds or even minutes in production because of this behaviour... wow, that is a **massive** footgun. No blame on websockets though.

**Comment by mishas** [comment]
@lgrahl,

Please also note that another difference between stream and websockets, is that websockets have internal (enabled by default) pings, which streams don't have.
Furthermore, if I don't control both sides of the connection - I not necessarily control the interval of said pings.

This whole ticket started from me not understanding why pings are not returned, while I do everything correctly / asynchronously.

With that said, I'm ok with the decision that there's nothing that can be done to fix the code, as long as a warning on `send()` is added to the documentation.

Please also note that the documentation for `send()` of async generators is just wrong. It says that it will yield control, but in fact it doesn't (see comment above)
Moreover, while I can manually add an `await asyncio.sleep(0)` after each `ws.send()` of strings, there's no simple place to add `await asyncio.sleep(0)` if I'm sending an async generator, which makes sending of such generators very dangerous.

**Comment by aaugustin** [comment]
> Indeed, there only two situations where send() yields control to the event loop:

OK, the wording isn't ideal. I believe my intent was "send() will no yield control except is these two situations, where it may yield control".

**Comment by aaugustin** [comment]
> Please also note that another difference between stream and websockets, is that websockets have internal (enabled by default) pings, which streams don't have.

Having pings enabled by default or not is a binary choice, so it's hard to please everyone. There was a lot of demand for detecting when the other side of the connection drops.

> Please also note that the documentation for `send()` of async generators is just wrong. It says that it will yield control, but in fact it doesn't (see comment above)

I will fix the documentation to explain there is no guarantee of yielding control.

> Moreover, while I can manually add an `await asyncio.sleep(0)` after each `ws.send()` of strings, there's no simple place to add `await asyncio.sleep(0)` if I'm sending an async generator, which makes sending of such generators very dangerous.

That's one way to look at the situation.

Another way is — what's the point of using an async generator rather than a (non-async) generator if the async generator doesn't yield control between values it generates?

---- 

Please be aware that you're coming across as very confident ("I do everything correctly") and quite combative. However, I don't think this level of confidence is completely warranted.

We still don't know how large your TCP send buffers are and whether your 1MB messages fit in there or not, which affects whether drain() yields control.

If we step back, you mention CPU-bound workloads and saturating write buffers, which are less than ideal conditions for a cooperative multitasking I/O framework. Perhaps you'd be better off with preemptive multitasking (threads).

**Comment by aaugustin** [comment]
> Moreover, while I can manually add an await asyncio.sleep(0) after each ws.send() of strings, there's no simple place to add await asyncio.sleep(0) if I'm sending an async generator, which makes sending of such generators very dangerous.

This is an example of over-confidence. You say there's no way, but actually it's just four lines of code to turn an async generator that doesn't yield control — I still don't understand the point of such an async generator — into one that does:

```python
async def yield_between_values(async_gen):
    async for value in async_gen:
        yield value
        await asyncio.sleep(0)
```

Please don't spread FUD like "sending async generators is very dangerous".

**Comment by mishas** [comment]
@aaugustin,

The "I do everything correctly" was a typo. It was suppose to be "while I **did** everything correctly", and I meant in the context of that specific function.
I'm confident to say so, as I did debug it for hours to find out if maybe I missed anything before opening this ticket.

I also think you should start by look at yourself before pointing out others' confidence or combativeness (e.g. "I don't remember ever making guarantees about yielding control to the event loop" - which is clearly in the documentation, or the whole "Here's my theory" answer, which clearly was written without fully reading or thinking about the ticket and the information that was already there).

To me it felt like you've read half of the ticket, assumed I don't know much about asyncio, and didn't even give me the benefit of the doubt.

Here's another example of over-confidence for you:
> I still don't understand the point of such an async generator

Here're two examples:
```python
async def get_data():
    data = await get_actual_data_from_db()  # this is what takes time, but yields control
    for item in data:
        yield json.dumps(item)  # For large items it does take time. Even if you use orjson or ujson.

async def another_example():
    async for page in get_data_from_another_server():  # this does yield control, and takes time per page.
        for item in page:  # pages can be very large.
            yield format_item(item)  # again - takes a bit of time.
```
Maybe in my example on top I should've done `range(2000)` and `time.sleep(0.01)`, or `json.dumps(<some data here>)` to make it clearer, but I believed that smaller numbers for the range will make this example easier to understand.

---

Now to an answer your "spread FUD" comment (which is also a bit derogative),

Yes, I know I can wrap all my async generators in something that does `await asyncio.sleep(0)` between values, but I hardly believe it's going to make nice, readable code.

I do think that the FUD is warranted. For me, it crashed a production system, so being surprised that it's not returning pings (by not releasing control), is a big deal.

---

I really appreciate the websockets project and I'm truly thankful to you, and others, who wrote it and made it available.
With that said, I do not appreciate being schooled by random people on the internet on the levels of my confidence, and whether it's warranted or not.

**Comment by aaugustin** [comment]
Since this discussion was getting very long, I tried to extract the reasoning in a new ticket. I hope it will help us agree on where we disagree:

- I think the problem is in how you generate the data
- You think the problem is in how websockets sends it

It's fine if we remain in disagreement here :-)

**Comment by aaugustin** [comment]
Slightly more detailed proposal to agree on where we disagree:

- I think your data generation process should yield because it's taking a long time
- You think websockets should yield because it does I/O

The reason for yielding on I/O is because I/O can be slow, especially if you're waiting for data. But writing to the send buffers is fast, and therefore doesn't require yielding.

The real reason for yielding is when something is slow.
